sam.c

```diff
@@ -1924,13 +1938,655 @@ int sam_parse1(kstring_t *s, sam_hdr_t *h, bam1_t *b)
+                cp--;
+
+            // entire buffer is part of a single line
+            if (cp == l->data) {
+                line_frag = l->data_size;
+                char *rp = realloc(l->data, l->alloc * 2);
+                if (!rp) {
+                    sam_state_err(fd, ENOMEM);
+                    goto err;
+                }
+                l->alloc *= 2;
+                l->data = rp;
+                assert(l->alloc >= l->data_size);
+                assert(l->alloc >= line_frag);
+                assert(l->alloc >= l->alloc - line_frag);
+                goto longer_line;
+            }
+            cp++;
+
+            // line holds the remainder of our line.
+            if (ks_resize(&line, cp_end - cp) < 0) {
+                sam_state_err(fd, ENOMEM);
+                goto err;
+            }
+            memcpy(line.s, cp, cp_end - cp);
+            line_frag = cp_end - cp;
+            l->data_size = l->alloc - line_frag;
+        } else {
+            // out of buffer
+            line_frag = 0;
+        }
+
+        l->serial = fd->serial++;
+        //fprintf(stderr, "Dispatching %p, %d bytes, serial %d\n", l, l->data_size, l->serial);
+        if (hts_tpool_dispatch(fd->p, fd->q, sam_parse_worker, l) < 0) {
+            sam_state_err(fd, ENOMEM);
+            goto err;
+        }
+        l = NULL;  // Now "owned" by sam_parse_worker()
+    }
+
+    if (hts_tpool_dispatch(fd->p, fd->q, sam_parse_eof, NULL) < 0) {
+        sam_state_err(fd, ENOMEM);
+        goto err;
+    }
+
+    // At EOF, wait for close request.
+    // (In future if we add support for seek, this is where we need to catch it.)
+    for (;;) {
+        pthread_mutex_lock(&fd->command_m);
+        if (fd->command == SAM_NONE)
+            pthread_cond_wait(&fd->command_c, &fd->command_m);
+        switch (fd->command) {
+        case SAM_CLOSE:
+            pthread_cond_signal(&fd->command_c);
+            pthread_mutex_unlock(&fd->command_m);
+            hts_tpool_process_destroy(fd->q);
+            fd->q = NULL;
+            goto tidyup;
+
+        default:
+            pthread_mutex_unlock(&fd->command_m);
+            break;
+        }
+    }
+
+ tidyup:
+    if (l) {
+        pthread_mutex_lock(&fd->lines_m);
+        l->next = fd->lines;
+        fd->lines = l;
+        pthread_mutex_unlock(&fd->lines_m);
+    }
+    free(line.s);
+
+    return NULL;
+
+ err:
+    hts_tpool_process_destroy(fd->q);
+    fd->q = NULL;
+    goto tidyup;
+}
+
+// Runs in its own thread.
+// Takes encoded blocks of SAM off the thread results queue and writes them
+// to our output stream.
+static void *sam_dispatcher_write(void *vp) {
+    htsFile *fp = vp;
+    SAM_state *fd = fp->state;
+    hts_tpool_result *r;
+
+    // Iterates until result queue is shutdown, where it returns NULL.
+    while ((r = hts_tpool_next_result_wait(fd->q))) {
+        sp_lines *gl = (sp_lines *)hts_tpool_result_data(r);
+        if (!gl)
+            continue;
```

Comment:
Shouldn't it give up here, as if `gl == NULL`, `sam_format_worker()` has likely failed?

(Aside: It might make this easier to follow if the `hts_tpool_next_result_wait()` calls had comments to say which function(s) the result may have come from.)

Reply:
I honestly cannot remember why I made it continue on empty results now, as it's lost on the mists of time.  That dates back to the original PR from last year.

I'll try changing it and see if anything breaks!  It's not obvious why it would be beneficial to continue.